ci_scripts/run_tests.sh

```diff
@@ -1,10 +1,98 @@
-set -euf -o pipefail
+#!/bin/bash
 
-carthage bootstrap --platform ios --configuration Release --no-use-binaries
-cd Example; carthage bootstrap --platform ios; cd ..
+function info {
+  echo "[$(basename "${0}")] [INFO] ${1}"
+}
 
-gem install xcpretty --no-ri --no-rdoc
-xcodebuild clean build build-for-testing -workspace Stripe.xcworkspace -scheme "StripeiOS" -configuration Debug -sdk iphonesimulator -destination 'platform=iOS Simulator,name=iPhone 6,OS=11.2' | xcpretty -c
-xcodebuild test-without-building -workspace Stripe.xcworkspace -scheme "StripeiOS" -configuration Debug -sdk iphonesimulator -destination 'platform=iOS Simulator,name=iPhone 6,OS=11.2' | xcpretty -c
-xcodebuild build -workspace Stripe.xcworkspace -scheme "Standard Integration (Swift)" -sdk iphonesimulator -destination 'platform=iOS Simulator,name=iPhone 6,OS=11.2'  | xcpretty -c
-xcodebuild build -workspace Stripe.xcworkspace -scheme "Custom Integration (ObjC)" -sdk iphonesimulator -destination 'platform=iOS Simulator,name=iPhone 6,OS=11.2' | xcpretty -c
+function die {
+  echo "[$(basename "${0}")] [ERROR] ${1}"
+  exit 1
+}
+
+# Verify xcpretty is installed
+if ! command -v xcpretty > /dev/null; then
+  if [[ "${CI}" != "true" ]]; then
+    die "Please install xcpretty: https://github.com/supermarin/xcpretty#installation"
+  fi
+
+  info "Installing xcpretty..."
+  gem install xcpretty --no-ri --no-rdoc || die "Executing \`gem install xcpretty\` failed"
+fi
+
+# Install test dependencies
+info "Installing test dependencies..."
+
+carthage bootstrap --platform iOS --configuration Release --no-use-binaries
+carthage_exit_code="$?"
+
+if [[ "${carthage_exit_code}" != 0 ]]; then
+  die "Executing carthage failed with status code: ${carthage_exit_code}"
+fi
+
+# Execute tests (iPhone 6 @ iOS 11.2)
+info "Executing tests (iPhone 6 @ iOS 11.2)..."
+
+xcodebuild clean test \
```

Comment:
Why did you combine the build & test steps? They were intentionally separated last year: https://github.com/stripe/stripe-ios/commit/eb8009acf315e4b5c5b7fdfe2cba46fde9236ec1

I know I've run into this problem in the past (and have the radar to prove it), but I don't know if it's still an issue in Xcode 9.

Reply:
i combined it to reduce code duplication, and it seems to work? i think if we need it again in the future, we should leave the reasoning as a comment in the code because a bum like me is going to miss the commit blame